### PR TITLE
Drop test recipe

### PR DESCRIPTION
**Pull request by mnabian**
# PhysicsNeMo Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: examples/structural_mechanics/drop_test/vtu_reader.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    for idx, name in enumerate(disp_names):
+        disp = np.asarray(mesh.point_data[name])
+        if disp.ndim != 2 or disp.shape[1] != 3:
+            raise ValueError(
+                f"Point-data array '{name}' must be a 3-component vector (got shape {disp.shape})."
+            )
+        if idx == 0:
+            pos_list.append(coords)
+        else:
+            pos_list.append(coords + disp)
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/drop_test/vtu_reader.py]: Fixed.

File: examples/structural_mechanics/drop_test/train.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+        else:
+            self.optimizer = torch.optim.Adam(
+                self.model.parameters(), lr=cfg.training.start_lr, fused=True
+            )
+        logger0.info(f"Using {self.optimizer.__class__.__name__} optimizer")
+
+        self.scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
+            self.optimizer, T_max=cfg.training.epochs, eta_min=cfg.training.end_lr
+        )
+        self.scaler = GradScaler("cuda", enabled=self.amp)
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/drop_test/train.py:216]: Fixed.

File: examples/structural_mechanics/drop_test/rollout.py [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+        """
+        device = sample.node_features["coords"].device
+        outputs: list[torch.Tensor] = []
+        for t in range(self.rollout_steps):
+            time = torch.tensor(t / self.rollout_steps, device=device)
+            sample.node_features["time"] = time
+            y_t2 = self._forward(sample, data_stats)
+            outputs.append(y_t2)
+
+        return torch.stack(outputs, dim=0).transpose(0, 1)  # [N,T,3]
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/drop_test/rollout.py]: Fixed.

File: examples/structural_mechanics/drop_test/datapipe.py [hunk 4]
```diff
@@ -0,0 +1,742 @@
...
+                        continue
+                    mu = self.dynamic_target_stats[f"{k}_mean"].flatten()
+                    std = self.dynamic_target_stats[f"{k}_std"].flatten()
+                    ts = self.target_series_data[i][k]
+                    if ts.ndim == 2:
+                        ts = ts.unsqueeze(-1)
+                    # ts: [T, N, C], mu/std: [C]
+                    self.target_series_data[i][k] = (
+                        ts - mu.view(1, 1, -1)
+                    ) / (std.view(1, 1, -1) + EPS)
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/drop_test/datapipe.py]: Fixed.

[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/drop_test/datapipe.py]: load_json returns tensors, so .flatten() works.

File: examples/structural_mechanics/drop_test/datapipe.py [hunk 5]
```diff
@@ -0,0 +1,742 @@
...
+                        "or targets will remain unnormalized."
+                    )
+
+        # Normalize trajectories and features
+        for i in range(self.num_samples):
+            self.mesh_pos_seq[i] = self._normalize_node_tensor(
+                self.mesh_pos_seq[i],
+                self.node_stats["pos_mean"],
+                self.node_stats["pos_std"],
+            )
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/drop_test/datapipe.py:369]: physicsnemo.datapipes.gnn.utils.load_json (line 98) explicitly wraps every loaded value in torch.tensor(v, dtype=torch.float); its return type is Dict[str, torch.Tensor] and its docstring matches. .view() on the loaded values works correctly. No change.

File: examples/structural_mechanics/drop_test/conf/model/figconvunet_one_shot.yaml [hunk 6]
```diff
@@ -0,0 +1,59 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 - 2026 NVIDIA CORPORATION & AFFILIATES.
```
Comment (RishikeshRanade) [inline examples/structural_mechanics/drop_test/conf/model/figconvunet_one_shot.yaml:1]: Do you think crash and droptest recipes can share the same configs? If so, it would make sense to have a unified recipe, right?

File: examples/structural_mechanics/drop_test/datapipe.py [hunk 7]
```diff
@@ -0,0 +1,754 @@
...
+from physicsnemo.utils.logging import PythonLogger
+
+NODE_STATS_FILE = "node_stats.json"
+FEATURE_STATS_FILE = "feature_stats.json"
+EDGE_STATS_FILE = "edge_stats.json"
+DYNAMIC_TARGET_STATS_FILE = "dynamic_target_stats.json"
+EPS = 1e-8  # numerical stability for std
+
+
+class SimSample:
```
Comment (RishikeshRanade) [inline examples/structural_mechanics/drop_test/datapipe.py:38]: If this class is common across multiple recipes, should we move it into the framework?

File: examples/structural_mechanics/drop_test/datapipe.py [hunk 8]
```diff
@@ -0,0 +1,754 @@
...
+        )
+        ts = (
+            ""
+            if self.target_series is None
+            else f", target_series={list(self.target_series.keys())}"
+        )
+        return f"SimSample(N={n}, keys={list(self.node_features.keys())}, Din={din}, Dout={dout}, E={e}{gf}{ts})"
+
+
+class DropTestBaseDataset:
```
Comment (RishikeshRanade) [inline examples/structural_mechanics/drop_test/datapipe.py:114]: Same comment as above. Are these similar across recipes? Maybe we have a base class in physicsnemo.datapipes.cae and inherit from that here.

File: examples/structural_mechanics/drop_test/rollout.py [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+                local_positions=coords.unsqueeze(0),
+                global_embedding=global_emb,
+            )
+            .squeeze(0)
+        )
+        pred = _oneshot_add_coords(_oneshot_output(raw, N, T, Fo), coords)
+        return pred
+
+
+class TransolverOneShot(Transolver):
```
Comment (RishikeshRanade) [inline examples/structural_mechanics/drop_test/rollout.py:120]: Should rollout schemes be part of the framework as they will be shared not only across crash recipes but other applications?

**Review by RishikeshRanade — APPROVED** [review]

File: examples/structural_mechanics/drop_test/conf/model/figconvunet_one_shot.yaml (hunk 6 above)
Comment (mnabian) [reply examples/structural_mechanics/drop_test/conf/model/figconvunet_one_shot.yaml:1]: We discussed this with Ram, and decided to keep the drop test as a separate recipe for more flexibility. Later, we can consider unifying it with the crash recipe, depending on how much overlap there is. Also, we expect the drop test recipe to undergo major changes, and by keeping it separate, we make sure we are not introducing unwanted changes to the crash recipe.

File: examples/structural_mechanics/drop_test/rollout.py (hunk 9 above)
Comment (mnabian) [reply examples/structural_mechanics/drop_test/rollout.py:120]: Yes I agree. We can move these to the core code in a separate PR.

File: examples/structural_mechanics/drop_test/datapipe.py (hunk 8 above)
Comment (mnabian) [reply examples/structural_mechanics/drop_test/datapipe.py:114]: Please see my comment above.

File: examples/structural_mechanics/drop_test/datapipe.py (hunk 7 above)
Comment (mnabian) [reply examples/structural_mechanics/drop_test/datapipe.py:38]: Please see my comment above.
